File: yt_rw.r
# yt_rw
# state-space model for Georges Bank yellowtail flounder survey data
# want to compare population estimates with simple average and MARSS
# created 18 January 2017  Chris Legault
# last modified 19 January 2017

library(TMB)
library(Hmisc)

# compile cpp code and load dll
compile("yt_rw.cpp")
dyn.load(dynlib("yt_rw"))

# read data from file
input <- read.csv("yt.csv", header = TRUE)

log_obs_DFO <- log(input$DFO.B / 1000)  # log scale indices in 1000 mt
log_obs_spring <- log(input$Spring.B / 1000)
log_obs_fall <- log(input$Fall.B / 1000)
DFO_sd <- sqrt(log(input$DFO.CV^2 + 1)) # compute log scale survey sd from CV
spring_sd <- sqrt(log(input$Spring.CV^2 + 1))
fall_sd <- sqrt(log(input$Fall.CV^2 + 1))

n.obs <- length(log_obs_DFO) # assumes data in all years for all three surveys

# set up data and parameters
dat <- list(
  log_obs_DFO = log_obs_DFO,
  log_obs_spring = log_obs_spring,
  log_obs_fall = log_obs_fall,
  DFO_sd = DFO_sd,
  spring_sd = spring_sd,
  fall_sd = fall_sd
)

parameters <- list(
  logB = rep(0,n.obs),
  log_process_error = 0,
  log_obs_error_DFO = 0,
  log_obs_error_spring = 0,
  log_obs_error_fall = 0
)

# now estimate population, process error, and additional observation errors
obj <- MakeADFun(dat,
                 parameters,
                 DLL = "yt_rw", 
                 random = c("logB"), 
                 silent = TRUE)
opt <- nlminb(obj$par, 
              obj$fn, 
              obj$gr, 
              control = list(iter.max = 1000,
                             eval.max = 1000))
srep <- summary(sdreport(obj))
print(srep)

# get results in a format for plots
est.log.pop <- srep[rownames(srep) == "logB",]
est.process.error <- as.vector(round(srep[rownames(srep) == "process_error",],3))
est.obs.error.DFO <- as.vector(round(srep[rownames(srep) == "obs_error_DFO",],3))
est.obs.error.spring <- as.vector(round(srep[rownames(srep) == "obs_error_spring",],3))
est.obs.error.fall <- as.vector(round(srep[rownames(srep) == "obs_error_fall",],3))

TMB_fit_ev <- exp(est.log.pop[,1])
TMB_fit_hi <- exp(est.log.pop[,1] + 2*est.log.pop[,2])
TMB_fit_lo <- exp(est.log.pop[,1] - 2*est.log.pop[,2])
obs.DFO <- exp(log_obs_DFO)
obs.spring <- exp(log_obs_spring)
obs.fall <- exp(log_obs_fall)
avg.B <- (obs.DFO + obs.spring + obs.fall) / 3
my.range <- range(c(0, obs.DFO, obs.spring, obs.fall, hi, lo))

# make a pretty plot
plot(input$Year,obs.DFO,
     ylim = my.range,
     xlab = "Year of DFO and Spring Surveys",
     ylab = "Pop Biomass (1000 mt)", 
     pch = 0)
points(input$Year,obs.spring, pch=1)
points(input$Year,obs.fall, pch=2)
errbar(input$Year, ep, hi, lo, pch=3, col="red", errbar.col="red", add=TRUE)
lines(input$Year, avg.B, col = "blue")
legend('topleft', 
       legend = c("DFO", "Spring", "Fall", "est", "avg"),
       pch = c(0, 1, 2, 3, NA),
       lty = c(NA, NA, NA, NA, 1),
       col = c("black", "black", "black", "red", "blue"))
title(main = paste0("process error   = ",
                    est.process.error[1],
                    "\nadditional observation errors = ",
                    est.obs.error.DFO[1],
                    ", ",
                    est.obs.error.spring[1],
                    ", ",
                    est.obs.error.fall[1]), 
      outer=F)

cbind(input$Year, avg.B, ep)

File: yt_rw.cpp
#include <TMB.hpp>

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_VECTOR(log_obs_DFO);        // log scale values of expanded surveys
  DATA_VECTOR(log_obs_spring);
  DATA_VECTOR(log_obs_fall);
  DATA_VECTOR(DFO_sd);             // log scale standard deviations of surveys (computed from CVs in R code)
  DATA_VECTOR(spring_sd);
  DATA_VECTOR(fall_sd);
  
  PARAMETER_VECTOR(logB);          // population biomass on log scale
  PARAMETER(log_process_error);    // process error for population
  PARAMETER(log_obs_error_DFO);    // additional observation error for each survey (survey CV doesn't account for all variability)
  PARAMETER(log_obs_error_spring);
  PARAMETER(log_obs_error_fall);
  
  Type process_error = exp(log_process_error);
  Type obs_error_DFO = exp(log_obs_error_DFO);
  Type obs_error_spring = exp(log_obs_error_spring);
  Type obs_error_fall = exp(log_obs_error_fall);
  
  int n_obs = log_obs_DFO.size();  // number of observations
  
  Type nll=0;                      // negative log likelihood
  
  for(int y = 1; y < n_obs; y++){      // likelihood for state transitions
    //Type m = logB[y-1];
    nll -= dnorm(logB[y], logB[y-1], process_error, true);
  }
  
  for(int y = 0; y < n_obs; y++){      // likelihood for observations, note addition of obs_error and year-specific sd
    nll -= dnorm(log_obs_DFO[y],    logB[y], obs_error_DFO + DFO_sd[y],       true);
    nll -= dnorm(log_obs_spring[y], logB[y], obs_error_spring + spring_sd[y], true);
    nll -= dnorm(log_obs_fall[y],   logB[y], obs_error_fall + fall_sd[y],     true);
  }
  
  ADREPORT(process_error);
  ADREPORT(obs_error_DFO);
  ADREPORT(obs_error_spring);
  ADREPORT(obs_error_fall);
  
  return nll;
}
